### PR TITLE
fix the build error of #3940 

### PR DESCRIPTION
**Pull request by Acey9**
fix the error of  [#3940](https://github.com/celery/celery/pull/3940)

### CODE REVIEW

**Comment by georgepsarakis** [comment]
@Acey9 there are still two Travis CI jobs failing, one concerns PEP8 and the other the configuration documentation.

**Comment by Acey9** [comment]
very sorry, i fix the error about pep8. but i not found the error about documentation in my changed。 
return some error when I tried to compile the html document。but this is not my changed。
The "make html" return like this：
ubuntu:~/celery/docs# make html
sphinx-build -b html -d _build/doctrees   . _build/html
Running Sphinx v1.5.3
making output directory...
/usr/local/lib/python2.7/dist-packages/sphinx/util/requests.py:50: UserWarning: Some links may return broken results due to being unable to check the Server Name Indication (SNI) in the returned SSL cert against the hostname in the url requested. Recommended to install "requests[security]" as a dependency or upgrade to a python version with SNI support (Python 3 and Python 2.7.9+).
  'Some links may return broken results due to being unable to '
loading pickled environment... not yet created
loading intersphinx inventory from http://sqlalchemy.readthedocs.io/en/latest/objects.inv...
loading intersphinx inventory from http://docs.python.org/dev/objects.inv...
intersphinx inventory has moved: http://docs.python.org/dev/objects.inv -> https://docs.python.org/dev/objects.inv
loading intersphinx inventory from http://kazoo.readthedocs.io/en/latest/objects.inv...
loading intersphinx inventory from http://vine.readthedocs.io/en/latest/objects.inv...
loading intersphinx inventory from http://gevent.org/objects.inv...
intersphinx inventory has moved: http://gevent.org/objects.inv -> http://www.gevent.org/objects.inv
loading intersphinx inventory from http://django-celery.readthedocs.io/en/latest/objects.inv...
loading intersphinx inventory from http://tox.readthedocs.io/en/latest/objects.inv...
loading intersphinx inventory from http://pythonhosted.org/msgpack-python/objects.inv...
WARNING: intersphinx inventory u'http://pythonhosted.org/msgpack-python/objects.inv' not fetchable due to <class 'requests.exceptions.HTTPError'>: 404 Client Error: Not Found
loading intersphinx inventory from http://www.sphinx-doc.org/en/stable/objects.inv...
loading intersphinx inventory from http://amqp.readthedocs.io/en/latest/objects.inv...
loading intersphinx inventory from http://boto.readthedocs.io/en/latest/objects.inv...
loading intersphinx inventory from http://eventlet.net/doc/objects.inv...
loading intersphinx inventory from http://pyopenssl.readthedocs.io/en/stable/objects.inv...
loading intersphinx inventory from http://redis-py.readthedocs.io/en/latest/objects.inv...
loading intersphinx inventory from http://django.readthedocs.io/en/latest/objects.inv...
loading intersphinx inventory from http://kombu.readthedocs.io/en/master/objects.inv...
loading intersphinx inventory from http://cyme.readthedocs.io/en/latest/objects.inv...
loading intersphinx inventory from https://cyanide.readthedocs.io/en/latest/objects.inv...
loading intersphinx inventory from http://sendapatch.se/projects/pylibmc/objects.inv...
loading intersphinx inventory from http://basho.github.io/riak-python-client/objects.inv...
loading intersphinx inventory from http://flower.readthedocs.io/en/latest/objects.inv...
loading intersphinx inventory from http://nose.readthedocs.io/en/latest/objects.inv...
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 199 source files that are out of date
updating environment: 199 added, 0 changed, 0 removed
reading sources... [100%] whatsnew-4.0
/root/celery/docs/userguide/tasks.rst:1576: ERROR: Unexpected indentation.
/root/celery/docs/userguide/tasks.rst:1582: ERROR: Unexpected indentation.
/root/celery/docs/userguide/tasks.rst:1586: ERROR: Unexpected indentation.
/root/celery/docs/userguide/tasks.rst:1590: ERROR: Unexpected indentation.
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
writing output... [100%] whatsnew-4.0
/root/celery/celery/backends/amqp.py:docstring of celery.backends.amqp.AMQPBackend.Producer:None: WARNING: more than one target found for cross-reference u'Exchange': celery.backends.rpc.RPCBackend.Exchange, celery.backends.amqp.AMQPBackend.Exchange
/root/celery/celery/backends/amqp.py:docstring of celery.backends.amqp.AMQPBackend.Producer.publish:None: WARNING: more than one target found for cross-reference u'Exchange': celery.backends.rpc.RPCBackend.Exchange, celery.backends.amqp.AMQPBackend.Exchange
/root/celery/celery/backends/rpc.py:docstring of celery.backends.rpc.RPCBackend.Producer:None: WARNING: more than one target found for cross-reference u'Exchange': celery.backends.rpc.RPCBackend.Exchange, celery.backends.amqp.AMQPBackend.Exchange
/root/celery/celery/backends/rpc.py:docstring of celery.backends.rpc.RPCBackend.Producer.publish:None: WARNING: more than one target found for cross-reference u'Exchange': celery.backends.rpc.RPCBackend.Exchange, celery.backends.amqp.AMQPBackend.Exchange
/root/celery/celery/app/base.py:docstring of celery.Celery:None: WARNING: more than one target found for cross-reference u'type': celery.worker.request.Request.type, celery.app.control.Control.Mailbox.type, celery.backends.amqp.AMQPBackend.Exchange.type, celery.utils.abstract.CallableSignature.type, celery.backends.rpc.RPCBackend.Exchange.type
/root/celery/celery/app/base.py:docstring of celery.Celery:None: WARNING: more than one target found for cross-reference u'type': celery.worker.request.Request.type, celery.app.control.Control.Mailbox.type, celery.backends.amqp.AMQPBackend.Exchange.type, celery.utils.abstract.CallableSignature.type, celery.backends.rpc.RPCBackend.Exchange.type
/root/celery/celery/app/base.py:docstring of celery.Celery:None: WARNING: more than one target found for cross-reference u'type': celery.worker.request.Request.type, celery.app.control.Control.Mailbox.type, celery.backends.amqp.AMQPBackend.Exchange.type, celery.utils.abstract.CallableSignature.type, celery.backends.rpc.RPCBackend.Exchange.type
/root/celery/celery/app/base.py:docstring of celery.Celery:None: WARNING: more than one target found for cross-reference u'type': celery.worker.request.Request.type, celery.app.control.Control.Mailbox.type, celery.backends.amqp.AMQPBackend.Exchange.type, celery.utils.abstract.CallableSignature.type, celery.backends.rpc.RPCBackend.Exchange.type
/root/celery/celery/app/base.py:docstring of celery.Celery:None: WARNING: more than one target found for cross-reference u'type': celery.worker.request.Request.type, celery.app.control.Control.Mailbox.type, celery.backends.amqp.AMQPBackend.Exchange.type, celery.utils.abstract.CallableSignature.type, celery.backends.rpc.RPCBackend.Exchange.type
/root/celery/celery/app/base.py:docstring of celery.Celery:None: WARNING: more than one target found for cross-reference u'type': celery.worker.request.Request.type, celery.app.control.Control.Mailbox.type, celery.backends.amqp.AMQPBackend.Exchange.type, celery.utils.abstract.CallableSignature.type, celery.backends.rpc.RPCBackend.Exchange.type
/root/celery/celery/app/base.py:docstring of celery.Celery:None: WARNING: more than one target found for cross-reference u'type': celery.worker.request.Request.type, celery.app.control.Control.Mailbox.type, celery.backends.amqp.AMQPBackend.Exchange.type, celery.utils.abstract.CallableSignature.type, celery.backends.rpc.RPCBackend.Exchange.type
/root/celery/celery/canvas.py:docstring of celery.Signature:None: WARNING: more than one target found for cross-reference u'Task': celery.app.task.Task, celery.events.state.Task, celery.events.state.State.Task, celery.Celery.Task, celery.backends.database.models.Task
/root/celery/celery/app/base.py:docstring of celery.app.default_app:None: WARNING: more than one target found for cross-reference u'type': celery.worker.request.Request.type, celery.app.control.Control.Mailbox.type, celery.backends.amqp.AMQPBackend.Exchange.type, celery.utils.abstract.CallableSignature.type, celery.backends.rpc.RPCBackend.Exchange.type
/root/celery/celery/app/base.py:docstring of celery.app.default_app:None: WARNING: more than one target found for cross-reference u'type': celery.worker.request.Request.type, celery.app.control.Control.Mailbox.type, celery.backends.amqp.AMQPBackend.Exchange.type, celery.utils.abstract.CallableSignature.type, celery.backends.rpc.RPCBackend.Exchange.type
/root/celery/celery/app/base.py:docstring of celery.app.default_app:None: WARNING: more than one target found for cross-reference u'type': celery.worker.request.Request.type, celery.app.control.Control.Mailbox.type, celery.backends.amqp.AMQPBackend.Exchange.type, celery.utils.abstract.CallableSignature.type, celery.backends.rpc.RPCBackend.Exchange.type
/root/celery/celery/app/base.py:docstring of celery.app.default_app:None: WARNING: more than one target found for cross-reference u'type': celery.worker.request.Request.type, celery.app.control.Control.Mailbox.type, celery.backends.amqp.AMQPBackend.Exchange.type, celery.utils.abstract.CallableSignature.type, celery.backends.rpc.RPCBackend.Exchange.type
/root/celery/celery/app/base.py:docstring of celery.app.default_app:None: WARNING: more than one target found for cross-reference u'type': celery.worker.request.Request.type, celery.app.control.Control.Mailbox.type, celery.backends.amqp.AMQPBackend.Exchange.type, celery.utils.abstract.CallableSignature.type, celery.backends.rpc.RPCBackend.Exchange.type
/root/celery/celery/app/base.py:docstring of celery.app.default_app:None: WARNING: more than one target found for cross-reference u'type': celery.worker.request.Request.type, celery.app.control.Control.Mailbox.type, celery.backends.amqp.AMQPBackend.Exchange.type, celery.utils.abstract.CallableSignature.type, celery.backends.rpc.RPCBackend.Exchange.type
/root/celery/celery/app/base.py:docstring of celery.app.default_app:None: WARNING: more than one target found for cross-reference u'type': celery.worker.request.Request.type, celery.app.control.Control.Mailbox.type, celery.backends.amqp.AMQPBackend.Exchange.type, celery.utils.abstract.CallableSignature.type, celery.backends.rpc.RPCBackend.Exchange.type
/root/celery/celery/result.py:docstring of celery.result.ResultSet:None: WARNING: more than one target found for cross-reference u'AsyncResult': celery.contrib.abortable.AbortableTask.AsyncResult, celery.Celery.AsyncResult, celery.app.task.Task.AsyncResult, celery.result.AsyncResult
/root/celery/celery/result.py:docstring of celery.result.GroupResult:None: WARNING: more than one target found for cross-reference u'AsyncResult': celery.contrib.abortable.AbortableTask.AsyncResult, celery.Celery.AsyncResult, celery.app.task.Task.AsyncResult, celery.result.AsyncResult
/root/celery/docs/userguide/monitoring.rst:291: WARNING: unknown option: flower --port
generating indices... genindex py-modindex
highlighting module code... [100%] celery.utils.iso8601
writing additional pages... search
copying images... [100%] userguide/../images/worker_graph_full.png
copying static files... done
copying extra files... done
dumping search index in English (code: en) ... done
dumping object inventory... done
build succeeded, 27 warnings.

Build finished. The HTML pages are in _build/html.

**Comment by Acey9** [comment]
sorry,i see, i will fix it

**Comment by Acey9** [comment]
I think i fix it all, please help me review it, thanks.

**Comment by Acey9** [comment]
Okay, I'll add a unit test later on

**Comment by georgepsarakis** [comment]
@Acey9 thank you! If you need any help please let me know.

**Comment by Acey9** [comment]
@georgepsarakis Yes,I need you help me.I have tried to add test case, but without success, can you tell me how?

**Comment by georgepsarakis** [comment]
@Acey9 a good example for testing configuration parameters is the [test_RiakBackend.test_config_params](https://github.com/celery/celery/blob/master/t/unit/backends/test_riak.py#L77).

As far as the `_index` method is concerned, you can follow the example of `test_ElasticsearchBackend.test_get`:

```
    def test_index(self):
        x = ElasticsearchBackend(app=self.app)
        x.doc_type = 'test-doc-type'
        x._server = Mock()
        x._server.index = Mock()
        # the return value of elasticsearch.Elasticsearch.index
        expected_result = ...
        x._server.index.return_value = expected_result
        body = {"field1": "value1"}
        index_result = x._index(id=sentinel.task_id, body=body, kwarg1='test1')
        x._server.index.assert_called_once_with(
            id=sentinel.task_id,
            doc_type=x.doc_type,
            index=x.index,
            body=body,
            kwarg1='test1'
        )
```

What do you think? Does this make sense? Let me know if you need further clarifications :smiley: !

**Comment by Acey9** [comment]
@georgepsarakis I add the test like this:
`

    def test_index(self):
        x = ElasticsearchBackend(app=self.app)
        x.doc_type = 'test-doc-type'
        x._server = Mock()
        x._server.index = Mock()
        expected_result = dict(_id=sentinel.task_id, _source={'result': sentinel.result})
        x._server.index.return_value = expected_result

        body = {"field1": "value1"}
        index_result = x._index(id=sentinel.task_id, body=body, kwarg1='test1')
        x._server.index.assert_called_once_with(
            id=sentinel.task_id,
            doc_type=x.doc_type,
            index=x.index,
            body=body,
            kwarg1='test1'
        )
    def test_config_params(self):
        self.app.conf.elasticsearch_max_retries = 10
        self.app.conf.elasticsearch_timeout = 20.0
        self.app.conf.elasticsearch_retry_on_timeout = True

        assert self.backend.es_max_retries == 10
        assert self.backend.es_timeout == 20.0
        assert self.backend.es_retry_on_timeout == True`


When I runing the test and did not succeed. The result is as follows:

~]$ py.test -xv celery/t/unit/backends/test_elasticsearch.py

celery/t/unit/backends/test_elasticsearch.py::test_ElasticsearchBackend::test_index PASSED
celery/t/unit/backends/test_elasticsearch.py::test_ElasticsearchBackend::test_config_params FAILED

self = <t.unit.backends.test_elasticsearch.test_ElasticsearchBackend instance at 0x7f176ef81b48>

#>       assert self.backend.es_max_retries == 10
#E       AttributeError: 'ElasticsearchBackend' object has no attribute 'es_max_retries'

I do not know why，do you have any ideas? test_index is ok？thinks

**Comment by georgepsarakis** [comment]
@Acey9 could you please add the test cases in this branch? It will be easier to review.

**Comment by Acey9** [comment]
@georgepsarakis  I add the test cases in this branch, please help me review it, thanks.

File: t/unit/backends/test_elasticsearch.py [hunk 1]
```diff
@@ -80,3 +80,32 @@ def test_backend_params_by_url(self):
...
+        )
+
+    def test_config_params(self):
+        self.app.conf.elasticsearch_max_retries = 10
+        self.app.conf.elasticsearch_timeout = 20.0
+        self.app.conf.elasticsearch_retry_on_timeout = True
+
+        assert self.backend.es_max_retries == 10
+        assert self.backend.es_timeout == 20.0
+        assert self.backend.es_retry_on_timeout == True
```
Comment (georgepsarakis) [inline t/unit/backends/test_elasticsearch.py]: It is generally more accurate to use object identity test for `True`, `False`, `None`:

```
assert self.backend.es_retry_on_timeout is True
```

File: t/unit/backends/test_elasticsearch.py [hunk 2]
```diff
@@ -80,3 +80,32 @@ def test_backend_params_by_url(self):
...
+        x.doc_type = 'test-doc-type'
+        x._server = Mock()
+        x._server.index = Mock()
+        expected_result = dict(
+            _id=sentinel.task_id,
+            _source={'result': sentinel.result}
+        )
+        x._server.index.return_value = expected_result
+
+        body = {"field1": "value1"}
```
Comment (georgepsarakis) [inline t/unit/backends/test_elasticsearch.py]: The test is failing because the `ElasticsearchBackend._index` method has not been called.

```
result = x._index(...)
assert result == expected_result
```

File: t/unit/backends/test_elasticsearch.py [hunk 3]
```diff
@@ -80,3 +80,33 @@ def test_backend_params_by_url(self):
...
+        x._index(id=sentinel.task_id, body=body, kwarg1='test1')
+        x._server.index.assert_called_once_with(
+            id=sentinel.task_id,
```
Comment (georgepsarakis) [inline t/unit/backends/test_elasticsearch.py]: I think you should convert this object to string in order to take into account this [change](https://github.com/celery/celery/pull/3961/files#diff-731afbc33e1f1197b570c2d5506600d2R109), so that the assertion works:

```
x._server.index.assert_called_once_with(
    id=string(sentinel.task_id),
```

File: t/unit/backends/test_elasticsearch.py [hunk 4]
```diff
@@ -80,3 +80,33 @@ def test_backend_params_by_url(self):
...
+            index=x.index,
+            body=body,
+            kwarg1='test1'
+        )
+
+    def test_config_params(self):
+        self.app.conf.elasticsearch_max_retries = 10
+        self.app.conf.elasticsearch_timeout = 20.0
+        self.app.conf.elasticsearch_retry_on_timeout = True
+
```
Comment (georgepsarakis) [inline t/unit/backends/test_elasticsearch.py]: Since you are updating the configuration and these parameters are set in the constructor, you need to re-initialize the backend instance, so that it re-reads the configuration:

```
self.backend = ElasticsearchBackend(app=self.app)
```

This should resolve the build job errors.

**Comment by Acey9** [comment]
@georgepsarakis  All is passed，thank you very very much for your help.

**Comment by georgepsarakis** [comment]
Great! Thank you for your efforts on this. Feel free to add yourself to the [contributors list](https://github.com/celery/celery/blob/master/CONTRIBUTORS.txt).
